daml-assistant/exe/DA/Daml/Assistant.hs

```diff
@@ -239,12 +259,60 @@ dispatch env path args = do
     requiredIO "Failed to spawn command subprocess." $
         runProcess_ (setEnv dispatchEnv $ proc path args)
 
-displayErrors :: IO () -> IO ()
-displayErrors m = m `catches`
-    [ Handler $ \ (e :: AssistantError) -> do
-        hPutStrLn stderr (displayException e)
-        exitFailure
-    , Handler $ \ (e :: ConfigError) -> do
-        hPutStrLn stderr (displayException e)
-        exitFailure
+handleErrors :: forall t. L.Handle IO -> IO t -> IO t
+handleErrors logger m = m `catches`
+    [ Handler (go . displayException @AssistantError)
+    , Handler (go . displayException @ConfigError)
     ]
+  where
+    go :: String -> IO t
+    go err = do
+        hPutStrLn stderr err
+        L.logJson logger L.Error $ mkLogTable
+            [ ("event", "error")
+            , ("message", A.String (T.pack err))
+            ]
+        exitFailure
+
+withLogger :: DamlPath -> (L.Handle IO -> IO ()) -> IO ()
+withLogger damlPath k = do
+    let logOfInterest prio = prio `elem` [L.Telemetry, L.Warning, L.Error]
+        gcpConfig = L.GCPConfig
+            { gcpConfigTag = "assistant"
+            , gcpConfigDamlPath = Just (unwrapDamlPath damlPath)
+            }
+
+        optedOutPath = unwrapDamlPath damlPath </> ".opted_out"
+
+    isOptedOut <- doesPathExist optedOutPath
+    if isOptedOut
+        then
+            k L.makeNopHandle
+        else do
+            L.withGcpLogger gcpConfig logOfInterest L.makeNopHandle $ \gcpState logger -> do
+                L.logMetaData gcpState
+                k logger
+
+-- | Get the arguments to `daml` and anonimize all but the first.
+-- That way, the daml command doesn't get accidentally anonimized.
+anonimizeArgs :: IO [T.Text]
+anonimizeArgs = do
+    args <- getArgs
+    case args of
+        [] -> pure []
+        argsHead : argsTail -> do
+            argsTail' <- concatMapM (anonimizeArg . T.pack) argsTail
+            pure (T.pack argsHead : argsTail')
+
+-- | Anonimize an argument to `daml`.
+anonimizeArg :: T.Text -> IO [T.Text]
```

Comment:
Nice, I like that anonimization strategy! Good compromise between not leaking sensitive data while giving us a good deal of information.

Reply:
I'd be more comfortable if we were a bit more aggressive here. I suppose what we're interested in are the commands the flags people are using. If we only keep stuff that matches the regular expression `[a-zA-Z0-9\-_]*`, we should learn enough.

Reply:
@hurryabit That seems reasonable to me. Should we still filter out file names though?

Reply:
Well, I added the extra character set test without removing the filter on existing file paths.

Reply:
I think having both is good.

Reply:
@cocreature & I decided to make it even more aggressive: If it isn't a flag, or on a whitelist (which I built by going manually through all the commands, and needs to be updated occasionally), it gets filtered out of the telemetry.